substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +54,63 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    class SpeculationReasonInvocationHandler implements InvocationHandler {
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return IsolatedSpeculationReasonEncoding.class.getDeclaredMethod(method.getName(), method.getParameterTypes()).invoke(args);
+        }
+    }
+
+    class SpeculationReasonEncodingSupplier implements Supplier<Object> {
+        private Object speculationReasonEncodingProxy;
+
+        SpeculationReasonEncodingSupplier(Object speculationReasonEncodingProxy) {
+            this.speculationReasonEncodingProxy = speculationReasonEncodingProxy;
+        }
+
+        @Override
+        public Object get() {
+            try {
+                return speculationReasonEncodingProxy.getClass().getDeclaredConstructor().newInstance();
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                e.printStackTrace();
```

Comment:
This is still an early version to get some feedback. Error checking will be improved once we agree on the core changes.